shell/platform/android/android_surface_gl.cc

```diff
@@ -176,7 +176,9 @@ AndroidContextGL* AndroidSurfaceGL::GLContextPtr() const {
 }
 
 std::unique_ptr<Surface> AndroidSurfaceGL::CreatePbufferSurface() {
-  onscreen_surface_ = GLContextPtr()->CreatePbufferSurface();
+  if (!onscreen_surface_ || !onscreen_surface_->IsValid()) {
```

Comment:
Maybe just an early return instead?  This needs a test too.

Reply:
If we early return, then'Rasterizer::DoMakeRasterSnapshot' will execute the following logic, which will change from hardware rasterization to software rasterization. The performance will be a little bit worse, but the logic should be correct. If you think an early return is more reasonable, I'm happy to tweak the code.

https://github.com/flutter/engine/blob/30f048de39236eaa4b7552a3d2d24feb1ada44fc/shell/common/rasterizer.cc#L292-L297

And about test, I'm not sure how to test this, would you mind giving me some suggestions ?  Thank you :) .

Reply:
I thought about it again, the early return will introduce a new problem. If `Rasterizer::DoMakeRasterSnapshot` is executed multiple times when the app is in the background. The first time there is no problem, and `AndroidSurfaceGL::on_screen_surface_` which size is 1x1 is created, but the second time, because `AndroidSurfaceGL::on_screen_surface_` already exists, it is returned early. As a result, `Rasterizer::DoMakeRasterSnapshot` has changed from hardware rasterization to software rasterization.

Reply:
I think the part I'm not clear on is how we'd get here.

Reply:
> I think the part I'm not clear on is how we'd get here.

Although I can't provide a reproducible sample, it appeared many times in my application, and I finally found the reason for the problem by printing the Log, and the problem was triggered by this line of code.
https://github.com/flutter/flutter/blob/019f03374c5b106c3fde05c8eee930cd26b8ffff/packages/flutter/lib/src/painting/shader_warm_up.dart#L95

Reply:
So about the scene in which the issue happened, I don't know if I made it clear or not. In theory, in the two release versions 2.5 and 2.8, if flutter is used normally, only 2.5 will encounter it. Because 2.5's `PaintingBinding.shaderWarmup` has a default value. So when the PaintingBinding is initialized, there will be a piece of logic that executes `Picture.toImage`.

`Picture.toImage` will finally call `Rasterizer.DoMakeRasterSnapshot` on raster thread. If at this moment we are in the situation where `AndroidSurfaceGL::on_screen_surface_` is set, but `Rasterizer::surface_` has not been set, the issue happens.  The above conclusions I found by logging in these methods to determine their order.

Reply:
Ahh interesting, that makes sense.

Reply:
I think this is a combination of a few things:

- The platform thread creates the onscreen surface
- It then calls `Shell::NotifyCreated`
- NotifyCreated fire and forgets the UI task runner, which might schedule an image decode (this is a recent change, from November or Decemberish)
- The rasterizer setup is called, but now there's a race where the UI thread might have asked the rasterizer to make a snapshot before it has been setup
- If the UI thread beats rasterizer setup, the rasterizer will say it doesn't have a surface to work with and ask for a snapshotting surface. The AndroidContextGL creates a pbuffer surface for this and sets it as the onscreen_surface_, which destroys the one we actually want.

Reply:
So what I think should happen here is the following:

- Rename `AndroidSurfaceGL::CreatePbufferSurface` to something like `AndroidSurfaceGL::GetOrCreateSnapshotSurface`.
- Have that renamed method only create the Pbuffer surface if `onscreen_surface_` is null/invalid. Otherwise, return `onscreen_surface_`.
- Write a test that exercises this, e.g. latch the raster task runner before `NotifyCreated` is called, run some Dart code that calls and completes `toImage`, unlatch the raster task runner, make sure the rasterizer surface is the right one.

Reply:
> Have that renamed method only create the Pbuffer surface if onscreen_surface_ is null/invalid. Otherwise, return onscreen_surface_.

What happens if the buffer isn't a Pbuffer at this time?  Does it make sense to destroy the existing surface and then create the actual Pbuffer?

Reply:
No. We just want to make sure that we have a GPU backed surface and don't end up in the software backend code.

Reply:
> So what I think should happen here is the following:
> 
> * Rename `AndroidSurfaceGL::CreatePbufferSurface` to something like `AndroidSurfaceGL::GetOrCreateSnapshotSurface`.
> * Have that renamed method only create the Pbuffer surface if `onscreen_surface_` is null/invalid. Otherwise, return `onscreen_surface_`.
> * Write a test that exercises this, e.g. latch the raster task runner before `NotifyCreated` is called, run some Dart code that calls and completes `toImage`, unlatch the raster task runner, make sure the rasterizer surface is the right one.

@dnfield 
We can't return `onscreen_surface_` because the type of the return value is `std::unique_ptr<Surface>`, but the type of `onscreen_surface_` is `std::unique_ptr<AndroidEGLSurface>`. So in either case we have to create a new `std::unique_ptr<Surface>`. so I renamed it to `AndroidSurfaceGL::CreateSnapshotSurface`. I also added test cases as you said in discord.